File: src/main/java/algo/backtrack/MatchsticksToSquare_473.java
package algo.backtrack;

import java.util.Arrays;

public class MatchsticksToSquare_473 {
    public static boolean makesquare(int[] matchsticks) {
        var len = matchsticks.length;
        var total = (int)Arrays.stream(matchsticks).asLongStream().sum();
        if (total % 4 != 0) {
            return false;
        }
        var edgeLen = total / 4;
        return backtrack(matchsticks, edgeLen, 4, edgeLen, 0, new int[5][1 << (len + 1)]);
    }

    private static boolean backtrack(final int[] matchsticks, final int edgeLen, int edgeNum, int leftLen, int bitmask, int[][] dp) {
        // we have matched all edges of the square
        if (leftLen == 0 && edgeNum == 1) {
            return true;
        }
        // bitmask is the binary representation of the state
        if (dp[edgeNum][bitmask] != 0) {
            return dp[edgeNum][bitmask] == 1;
        }
        if (leftLen == 0) {
            // matched one edge
            edgeNum--;
            leftLen = edgeLen;
        }
        var len = matchsticks.length;
        for (var i = 0; i < len; i++) {
            var pos = 1 << (len - i);
            if (matchsticks[i] > leftLen) {
                continue;
            }
            if ((bitmask & pos) != 0) {
                continue;
            }
            var mask = bitmask ^ pos;
            if (backtrack(matchsticks, edgeLen, edgeNum, leftLen - matchsticks[i], mask, dp)) {
                dp[edgeNum][mask] = 1;
                return true;
            }
            dp[edgeNum][mask] = -1;
        }
        return false;
    }

    public static void main(String[] args) {
        System.out.println(makesquare(new int[]{1,1,2,2,2}));
        System.out.println(makesquare(new int[]{3,3,3,3,4}));
        System.out.println(makesquare(new int[]{5,5,5,5,4,4,4,4,3,3,3,3}));
        System.out.println(makesquare(new int[]{16,8,8,8,5,1,16,3,11,1,11,12,20,6,6}));
        System.out.println(makesquare(new int[]{99,37,37,37,37,37,37,37,37,5}));
        System.out.println(makesquare(new int[]{5,5,5,5,4,4,4,4,3,3,3,3}));
    }
}

File: src/main/kotlin/algo/backtrack/MatchsticksToSquare_473.kt
package algo.backtrack

class KtMatchsticksToSquare_473 {
    fun makesquare(matchsticks: IntArray): Boolean {
        val total = matchsticks.sum()
        if (total % 4 != 0) {
            return false
        }
        val side = total / 4

        fun match(sticks: IntArray, sum: Int, sideNum: Int, bitmask: Int, dp: Array<IntArray>): Int {
            if (sum == total && sideNum == 4) {
                return 1
            }
            if (dp[sideNum][bitmask] != 0) {
                return dp[sideNum][bitmask]
            }
            val sideSum = sum % side
            for (i in sticks.indices) {
                val nextSum = sideSum + sticks[i]
                if (nextSum > side) {
                    continue
                }
                val pos = 1 shl (sticks.size - i)
                if (pos and bitmask != 0) {
                    continue
                }
                val nextSideNum = if (nextSum == side) sideNum + 1 else sideNum
                val res = match(sticks, sum + sticks[i], nextSideNum, pos xor bitmask, dp)
                dp[sideNum][bitmask] = res
                if (res == 1) {
                    break
                }
                dp[sideNum][bitmask] = -1
            }
            return dp[sideNum][bitmask]
        }

        val stateArray = Array(5) { IntArray(1 shl (matchsticks.size + 1)) {0} }
        return match(matchsticks, 0, 0, 0, stateArray) == 1
    }

}

fun main() {
    val matchSticks = KtMatchsticksToSquare_473()
    println(matchSticks.makesquare(intArrayOf(1,1,2,2,2)))
    println(matchSticks.makesquare(intArrayOf(3,3,3,3,4)))
    println(matchSticks.makesquare(intArrayOf(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15)))
    println(matchSticks.makesquare(intArrayOf(5969561,8742425,2513572,3352059,9084275,2194427,1017540,2324577,6810719,8936380,7868365,2755770,9954463,9912280,4713511)))
    println(matchSticks.makesquare(intArrayOf(5,5,5,5,16,4,4,4,4,4,3,3,3,3,4)))
}